madmom/ml/nn/layers.py

```diff
@@ -344,6 +344,144 @@ def activate(self, data):
+    ----------
+    weights_in_to_hidden_update : numpy array, shape ()
+        Weights.
+    weights_in_to_hidden_update : numpy array, shape ()
+        Weights.
+    bias : scalar or numpy array, shape ()
+        Bias.
+    activation_fn : numpy ufunc, optional
+        Activation function.
+
+    References
+    ----------
+    .. [1] Cho, Kyunghyun, et al,
+           "On the properties of neural machine translation: Encoder-decoder
+           approaches",
+           arXiv preprint arXiv:1409.1259 (2014).
+
+    """
+
+    def __init__(self, weights_in_to_hidden_update,
+                 weights_hid_to_hidden_update, bias, activation_fn=tanh):
+        self.weights_in_to_hidden_update = weights_in_to_hidden_update
+        self.weights_hid_to_hidden_update = weights_hid_to_hidden_update
+        self.bias = bias
+        self.activation_fn = activation_fn
+
+    def activate(self, data, reset_gate, prev_hid):
+        """
+        Activate the gate with the given input, reset_gate and the previous
+        output:
+        h = tanh(Wxh * xt + Whh * ht-1 + b)
+
+        Parameters
+        ----------
+        data : scalar or numpy array, shape ()
+            Input data for the cell.
+        reset_gate : scalar or numpy array, shape ()
+            Activation of the reset gate.
+        prev_hid : scalar or numpy array, shape ()
+            Output data of the previous time step.
+
+        Returns
+        -------
+        numpy array
+            Activations of the gate for this data.
+
+        """
+        # weight input and add bias
+        in_to_hidden_update = np.dot(
+            data, self.weights_in_to_hidden_update) + self.bias
+        # weight previous hidden state and reset gate
+        hid_to_hidden_update = reset_gate * np.dot(
+            prev_hid, self.weights_hid_to_hidden_update)
+        # apply activation function and return it
+        return self.activation_fn(in_to_hidden_update + hid_to_hidden_update)
+
+
+class GRULayer(Layer):
+    """
+    Recurrent network layer with Gated Recurrent Units (GRU) as proposed in
+    [1].
+
+    Parameters
+    ----------
+    reset_gate : Cell instance
+        Reset gate.
+    update_gate : Cell instance
+        Update gate.
+    gru_cell : GRUCell instance
+        Hidden update
+    hid_init : numpy array, optional
+        Initial state of hidden units.
+
+    References
+    ----------
+    .. [1] Cho, Kyunghyun, et al,
+           "On the properties of neural machine translation: Encoder-decoder
+           approaches",
+           arXiv preprint arXiv:1409.1259 (2014).
+
+    """
+
+    def __init__(self, reset_gate, update_gate, gru_cell, hid_init=None):
+        # init the gates
+        self.reset_gate = reset_gate
+        self.update_gate = update_gate
+        self.gru_cell = gru_cell
+        if hid_init is None:
+            self.hid_init = np.zeros(self.gru_cell.bias.size, dtype=NN_DTYPE)
+        else:
+            self.hid_init = hid_init.squeeze()
+
```

Comment:
I'm not sure if it is better to just init the variable with the given parameter and defer the check to the `activate()` method. Saved models become smaller if not used.